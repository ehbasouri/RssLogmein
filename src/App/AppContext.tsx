import React, {createContext} from 'react';

export interface AppContextType {
  feeds: any[];
  loading: boolean;
  feedDetails: any;
  fetchRss: (e: string, navigation: any) => void;
  addOrRemoveFromFavourites: (e: string, add: boolean) => void;
  favouriteFeeds: string[];
  onEntryPress: (e: string) => void;
}

interface AppContextProviderProps {
  children?: React.ReactNode;
  value: AppContextType;
}

export const AppContext = createContext<Partial<AppContextType>>({});

export function AppContextProvider({children, value}: AppContextProviderProps) {
  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}
