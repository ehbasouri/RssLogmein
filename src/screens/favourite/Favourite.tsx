import React from 'react';
import {Text, View} from 'react-native';

function Favourite() {
  return (
    <View>
      <Text>favourite</Text>
    </View>
  );
}

export default Favourite;
