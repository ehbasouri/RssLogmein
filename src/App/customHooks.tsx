import React, {useEffect, useState} from 'react';
import * as rssParser from 'react-native-rss-parser';
import {StorageManager} from '../lib/storage.service';
import {Linking} from 'react-native';
import {AppContextType} from './AppContext';

const FeedUrl = [
  'https://rss.nytimes.com/services/xml/rss/nyt/HomePage.xml',
  'https://www.nasa.gov/rss/dyn/breaking_news.rss',
  'https://feeds.npr.org/1001/rss.xml',
];

export function CustomHooks(): AppContextType {
  const [feeds, setFeeds] = useState<any[]>(FeedUrl);
  const [favouriteFeeds, setFavouriteFeeds] = useState<any[]>([]);
  const [feedDetails, setFeedDetails] = useState<any>({});
  const [loading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    getFavouriteList();
  }, []);

  async function fetchRss(url: string, navigation: any) {
    setLoading(true);
    fetch(url)
      .then(response => response.text())
      .then(responseData => rssParser.parse(responseData))
      .then(rss => {
        setFeedDetails(rss);
        navigation.navigate('entry', rss);
        setLoading(false);
      })
      .catch(err => {
        console.log('err : ', err);
        setLoading(false);
      });
  }

  async function getFavouriteList() {
    const favList: string[] = await StorageManager.getValue('favouriteFeeds');
    if (favList) {
      setFavouriteFeeds(favList);
    }
  }

  function checkInFavoriteFeeds(feed: string, isAddAction: boolean): string[] {
    if (favouriteFeeds?.length === 0) {
      return [feed];
    }
    const result = favouriteFeeds.includes(feed);
    if (!result) {
      return [feed, ...favouriteFeeds];
    }
    if (!isAddAction) {
      return favouriteFeeds.filter(element => element !== feed);
    }
    return favouriteFeeds;
  }

  function addOrRemoveFromFavourites(feed: string, add: boolean) {
    const newFavouriteList = checkInFavoriteFeeds(feed, add);
    setFavouriteFeeds(newFavouriteList);
    StorageManager.setValue('favouriteFeeds', newFavouriteList);
  }

  function onEntryPress(url: string): void {
    Linking.canOpenURL(url).then(supported => {
      if (supported) {
        Linking.openURL(url);
      } else {
        console.log("Don't know how to open URI: " + url);
      }
    });
  }

  return {
    feeds,
    loading,
    fetchRss,
    feedDetails,
    addOrRemoveFromFavourites,
    favouriteFeeds,
    onEntryPress,
  };
}
