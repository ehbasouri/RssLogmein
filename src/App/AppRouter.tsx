import * as React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {
  createNativeStackNavigator,
  NativeStackScreenProps,
} from '@react-navigation/native-stack';
import {ENTRY, FAVOURITE_LIST, FEED_LIST} from '../consts';
import FeedList from '../screens/feed/FeedList';
import Entry from '../screens/entry/Entry';
import Favourite from '../screens/favourite/Favourite';
import * as rssParser from 'react-native-rss-parser';

const NullHeader = () => null;

type FeedListParams = {
  feed_list: any;
  entry: rssParser.Feed;
  favourite_list: any;
};

export type FeedListNavigationProps = NativeStackScreenProps<
  FeedListParams,
  'feed_list'
>;
export type EntryNavigationProps = NativeStackScreenProps<
  FeedListParams,
  'entry'
>;
export type FavouriteNavigationProps = NativeStackScreenProps<
  FeedListParams,
  'entry'
>;

const Stack = createNativeStackNavigator();

export function AppStack() {
  return (
    <Stack.Navigator>
      <Stack.Screen
        options={{
          header: NullHeader,
        }}
        name={FEED_LIST}
        component={FeedList}
      />
      <Stack.Screen
        options={{
          header: NullHeader,
        }}
        name={ENTRY}
        component={Entry}
      />
      <Stack.Screen
        options={{
          header: NullHeader,
        }}
        name={FAVOURITE_LIST}
        component={Favourite}
      />
    </Stack.Navigator>
  );
}

function AppRouter(): JSX.Element {
  return (
    <NavigationContainer>
      <AppStack />
    </NavigationContainer>
  );
}

export default AppRouter;
