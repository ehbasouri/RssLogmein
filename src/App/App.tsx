import React from 'react';
import AppRouter from './AppRouter';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import reducers from '../redux/reducer';

export default function App(): JSX.Element {
  return (
    <Provider store={createStore(reducers, {})}>
      <AppRouter />
    </Provider>
  );
}
