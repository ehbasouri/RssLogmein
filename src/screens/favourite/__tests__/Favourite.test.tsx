import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import Favourite from '../Favourite';
import {AppContext} from '../../../App/AppContext';

describe('test suite', () => {
  test('Favourite test', async () => {
    const props: any = {
      favouriteFeeds: ['url'],
      navigation: {goBack: jest.fn()},
    };
    const {getByText, toJSON} = render(
      <AppContext.Provider value={props}>
        <Favourite {...props} />
      </AppContext.Provider>,
    );
    expect(getByText('url')).toBeTruthy();
    expect(getByText('Favourite List')).toBeTruthy();
    const button = getByText('BACK');
    expect(button).toBeTruthy();
    fireEvent.press(button);
    expect(props.navigation.goBack).toHaveBeenCalled();
    expect(toJSON()).toMatchSnapshot();
  });
});
