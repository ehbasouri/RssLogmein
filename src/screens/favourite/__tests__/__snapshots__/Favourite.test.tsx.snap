// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suite Favourite test 1`] = `
<View>
  <RCTScrollView
    ListHeaderComponent={
      <RssHeader
        onBackPress={[Function]}
        showBack={true}
        title="Favourite List"
      />
    }
    data={
      Array [
        "url",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
      ]
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 16,
              "paddingHorizontal": 6,
              "paddingTop": 50,
            }
          }
        >
          <Text
            onPress={[Function]}
            style={
              Object {
                "color": "blue",
                "flex": 1,
              }
            }
          >
            BACK
          </Text>
          <Text
            style={
              Object {
                "flex": 4,
                "fontSize": 18,
                "fontWeight": "bold",
                "width": 40,
              }
            }
          >
            Favourite List
          </Text>
          <Text
            style={
              Object {
                "color": "blue",
                "flex": 1,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#C1C1C1",
              "borderWidth": 0.5,
              "marginTop": 20,
              "opacity": 1,
              "padding": 20,
            }
          }
          testID="feed-item"
        >
          <Text
            testID="feed-title"
          >
            url
          </Text>
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="favourite-button"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                remove from Favourite
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
